File: Day278/palindromicPartitioning.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG{
    public static void main(String args[])throws IOException
    {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(in.readLine());
        while(t-- > 0){
            String str = in.readLine();
            Solution ob = new Solution();
            System.out.println(ob.palindromicPartition(str));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution{
    static int arr[];
    static int palindromicPartition(String str)
    {
        // code here
        int n = str.length();
        arr = new int[n];
        Arrays.fill(arr, -1);
        return pal(0, n, str)-1;
    }
    static int pal(int ind, int n, String str) {
        if (ind == n) {
            return 0;
        }
        if (arr[ind] != -1) {
            return arr[ind];
        }
        int min = Integer.MAX_VALUE;
        for (int i=ind; i<n;  i++) {
            if (isPal(str, ind, i)) {
                min = Math.min(min, 1+pal(i+1, n, str));
            }
        }
        return arr[ind] = min;
    }
    static boolean isPal(String str, int i, int j) {
        while (i<j) {
            if (str.charAt(i) != str.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: Day278/palindromicPartitioning.go
package main

import (
	"fmt"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func isPal(str string, i, j int) bool {
	for i < j {
		if str[i] != str[j] {
			return false
		}
		i++
		j--
	}
	return true
}

func palindromicPartition(str string) int {
	n := len(str)
	arr := make([]int, n)
	for i := range arr {
		arr[i] = -1
	}

	var pal func(ind, n int, str string) int
	pal = func(ind, n int, str string) int {
		if ind == n {
			return 0
		}
		if arr[ind] != -1 {
			return arr[ind]
		}
		minPartitions := 1 << 31 - 1
		for i := ind; i < n; i++ {
			if isPal(str, ind, i) {
				minPartitions = min(minPartitions, 1+pal(i+1, n, str))
			}
		}
		arr[ind] = minPartitions
		return minPartitions
	}

	return pal(0, n, str) - 1
}

func main() {
	var t int
	fmt.Scan(&t)
	for t > 0 {
		var str string
		fmt.Scan(&str)
		result := palindromicPartition(str)
		fmt.Println(result)
		t--
	}
}
